engine/server.go

```diff
@@ -0,0 +1,60 @@
+package engine
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"time"
+)
+
+type Option struct{}
+
+func DefaultOption() *Option {
+	return nil
+}
+
+func (o *Option) WithPingInterval(time.Duration) *Option { return o }
+func (o *Option) WithPingTimeout(time.Duration) *Option  { return o }
+func (o *Option) WithMaxBufferSize(int) *Option          { return o }
+
+type FrameType int
+
+const (
+	FrameBinary FrameType = iota
+	FrameText
+)
+
+type Server struct{}
+
+func New(*Option) *Server {
+	return nil
+}
+
+// OnXXX should be called before serving HTTP.
+func (s *Server) OnOpen(func(Context, *http.Request) error)    {}
+func (s *Server) OnUpgrade(func(Context, *http.Request) error) {}
```

Comment:
i support u too

Reply:
I tend not to do it. For polling transport, requests are not the same for each message, as well as upgrading. This design makes requests clearly that they are requests of Open/Upgrade actions.

If users want to keep some data in head, they could store them in session storage.

If putting a request in context, I tend to put the request with the current HTTP request. For websocket, it always the same, but not for polling.

And I still don't want to put a request in context, because there is no way to get POST request when polling. I hope to hide this behavior.